CMSgov/mint-app
# [EASI-2936] General Characterstics ExistingModel from Model Name to Foreign Key Reference

# EASI-2936

## Changes and Description

- Flyway Migration
  - adding new fk column
  - mapping model id by matching name to fk column
- Update model
- Update schema



## How to test this change

Testing is a little unique for this since we have to emulate migration.

* Drag the migration script into another folder temporarily to prevent execution
* Bring down the containers and run a fresh migration from an empty database
* Open the plan_general_characteristics table
* We need to add values for in the existing_model columns to test 4 things: mapping to a model plan name stored in the model_plan table, mapping to an external plan ID stored in the existing_models table, resolving the anomalous case where a name matches an entry in both tables in which case we default to the id stored in the model_plan table, and the case where no model name matched
* 1, matching to the local model_plan table: simply add the name of a model_plan table entry in the plan_general_characteristics table’s existing_model column
* 2, matching to the existing_model table: add the name of the desired existing_model entry into the plan_general_characteristics table
* 3, match against both: change the name of a model_plan entry to match an existing_model entry, then add that same name into the plan_general_characteristics table
* 4, enter nonsense for the existing_model column which matches no model plan name
* Drag the migration script back into the migrations folder
* Run migration again using scripts/dev db:migrate
* Open the db migration container and assert there are no errors
* Open the plan_general_characteristic table and assert that the expected columns have been added and all data mappings are handled appropriately 

## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
